src/tephra2.ts

```diff
@@ -2,122 +2,299 @@
+// classes or smaller disk cell size
+function memo<R, T extends (...args: any[]) => R>(f: T): T {
+  const memory = new Map<string, R>();
+
+  const memoized = (...args: any[]) => {
+      if (!memory.get(args.join())) {
+          memory.set(args.join(), f(...args));
+      }
+
+      return memory.get(args.join());
+  };
+
+  return memoized as T;
+}
+
+const squaredDistance = (p1: Ipoint, p2: Ipoint) => ((p2.x - p1.x) ** 2) + ((p2.y - p1.y) ** 2);
+
+// returns an array of {x,y} points forming the eruption grid centered at xvent, yvent
+// memoized as we will frequently be calling this with the same values
+export const getEruptionDiskGrid = memo((xVent, yVent, diskRadius, diskGridSize) => {
+  const radiusSquared = diskRadius ** 2;
+  const halfCellSize = diskGridSize / 2;
+  const center = {x: xVent, y: yVent};
+  const grid: Ipoint[] = [];
+  for (let i = 0; i < 2 * diskRadius; i += diskGridSize) {
+    for (let j = 0; j < 2 * diskRadius; j += diskGridSize) {
+      const cell = {
+        x: xVent - diskRadius + i + halfCellSize,
+        y: yVent - diskRadius + j + halfCellSize
+      };
+      if (squaredDistance(center, cell) < radiusSquared) {
+        grid.push(cell);
+      }
+    }
+  }
+  return grid;
+});
+
+const getDensity = memo((phi: number) => {
+  if (phi <= phiC) {
+    return rhoPMin;
+  } else if (phi >= phiF) {
+    return rhoPMax;
+  } else {
+    const proportion = ((rhoPMax - rhoPMin) * (phi - phiF)) / (phiC - phiF);    // linear proportion
+    return rhoPMax - proportion;
+  }
+});
+
+/** returns diameter in m */
+const getDiameter = (phi: number) => 2 ** (-1 * phi) * 10 ** -3;
+// returns phi given a diameter in mm
+const getPhi = (diameterMM: number) => (Math.log(1 / (125 * (diameterMM / 1000))) / Math.log(2)) - 3;  // solved for phi
+
+/**
+ * @param phi Phi class (particle size)
+ * @param colHeight Column height in meters
+ */
+export const getSettlingSpeed = memo((phi: number, colHeight: number) => {
+  const atmLayerThickness = 500;
+  if (colHeight < atmLayerThickness) {
+    console.error(`colHeight must be >= ${atmLayerThickness} (was called with ${colHeight})`);
+  }
+  const adjustedColHeight = Math.max(colHeight, atmLayerThickness);
+
+  let vAux = 0;
+  let vLayer = 0;
+
+  for (let layer = atmLayerThickness; layer < adjustedColHeight + atmLayerThickness; layer += atmLayerThickness) {
+    const elev = adjustedColHeight - layer;
+    const rhoA = rhoAS * Math.exp(-1 * elev / 8200);
+    const rho = getDensity(phi) - rhoA;
+
+    const diameter = getDiameter(phi);
+
+    const vL = (g * diameter ** 2 * rho) / (18 * airVisc);
+    const vI = diameter * ((4 * g ** 2 * rho ** 2) / (225 * airVisc * rhoA)) ** (1 / 3);
+    const vT = ((3.1 * rho * g * diameter) / (rhoA)) ** 0.5;
+
+    const reL = (diameter * rhoA * vL) / airVisc;
+    // const reI = (diameter * rhoA * vI) / airVisc;
+    const reT = (diameter * rhoA * vT) / airVisc;
```

Comment:
I assume that this came from Robert, and I can't claim to be able to validate this by looking :)  

Some of the constant values like 8200, 18, 31, and 255 aren't obvious to me for example.

That being said, I am not too concerned with understanding this.  Having the tests is enough for our purposes here.